File: CMPS101S18PA1/apint.java

public class apint{
    static final int MAXARRAY = 3000;
    int[] number = new int[MAXARRAY];
    String num;
    int numberlen;
    boolean negative = false;
    apint(){//default constructor, allocates memory for apint
        int[] number= new int[MAXARRAY];
    }
    apint(String num){//converts a string of ints to an apint that has a flag to determine if number is negative or not
        this.num=num;//intialize the string of numbers
        char c[] = num.toCharArray();
        int str= c.length-1;
        int[] number= new int[MAXARRAY];
        if(c[0]== '-'){//checks if the string is negative or not
            negative=true;
        }
        int counter =MAXARRAY-1;
        for(int j = str;j >=0;j--){//start at the back of the string
            if(c[j]== '-'){
                number[counter] = (c[j]-'-');//cant hold a '-' in an array of ints
            }else{
                number[counter]= (c[j] -'0');//converts character ascii value to a number
                counter--;
            }
        }
        
        numberlen=MAXARRAY-1-counter;
        this.number=number;//intialize the array of ints
        
    }
    apint(int a){//creates an apint from an int
        String str = String.valueOf(a);//same logic in this method as the constructor with the string
        num=str;
        
        char c[] = str.toCharArray();
        int strlen= c.length-1;
        int[] number= new int[MAXARRAY];
        if(c[0]== '-'){//checks if the string is negative or not
            negative=true;
        }
        int counter =MAXARRAY-1;
        for(int j = strlen;j >=0;j--){//start at the back of the string
            if(c[j]== '-'){
                number[counter] = (c[j]-'-');//cant hold a '-' in an array of ints
            }else{
                number[counter]= (c[j] -'0');//converts character ascii value to a number
                counter--;
            }
        }
        
        numberlen=MAXARRAY-1-counter;
        this.number=number;
    }
    apint(int[] a){//method is used in the factorial function to make a new apint from an array of ints
        
        number=a;
    }
    String print(apint a){//Print method to make the apint look pretty without all the leading zeros
        int[] str = a.getNumber();
        String s = "";//make an empty string
        int counter =0;//counter is used to determine the when the leadings 0s need to be removed
        String zero = "0";
        if(a.Negative()){
            s= s+"-";//add a negative sign to the front of empty string if negative
        }
        while(str[counter]==0){//checks ends when has passed all the leading zeros
            counter++;
            if (counter==MAXARRAY){//if made it to the max array then there is a 0 stored
                return zero;
            }
        }
        for(int i =0+counter;i<MAXARRAY; i++){//starts from counter to maxarray
            s=s+str[i];//add number to the back of string
        }
        return s;


    }
    int getNumLen(){//method to figure out the length of the number
        return numberlen;
    }
    String getNum(){//method to get the string in the class
        return num;
    }
    int[] getNumber(){//method to get the array of ints
        return number;
    }
    boolean Negative(){//determines if int is negative or positive
        return negative;
    }
    apint addition(apint first, apint second){//method to add 2 apints
        int[] a = first.getNumber();
        int[] b = second.getNumber();
        apint sum = new apint();//apint to store the running sum throughout the method
        int[] n = sum.getNumber();
        int carry=0;//used to determine what needs to be carried
        int total=0;//used as a temp to determine int needs to be stored in the array
        int i =MAXARRAY-1;//start at the back of the list to add
        while(i>=0){
            total=a[i]+b[i]+carry;//total is the addition of the 2 numbers including the carry
            n[i]= total%10;//if total is  15 or 25, n[i]=5
            total=total/10;//if total is 15 or 25, total 1,2
            carry= total%10;//carry is now the total mod 10 count
            i--;//decrement i
        }
        return sum;// return the apint of the number
    }
    int[] add(int[] a, int[] b){// method is the same as the addition method except has 
        int[] n = new int[MAXARRAY];//array of int to store the running sum throughout the method
        int carry=0;//used to determine what needs to be carried
        int total=0;//used as a temp to determine int needs to be stored in the array
        int i =MAXARRAY-1;//start at the back of the list to add
        while(i>=0){
            total=a[i]+b[i]+carry;//total is the addition of the 2 numbers including the carry
            n[i]= total%10;//if total is  15 or 25, n[i]=5
            total=total/10;//if total is 15 or 25, total 1,2
            carry= total%10;//carry is now the total mod 10 count
            i--;//decrement i
        }
        return n;//Return the array of ints
    }
    int[] subtract(int[] a, int[] b){//subracts 2 array of ints rather than apints becuase used in methods as a sub method
        int[] n = new int[MAXARRAY];
        int carry=0;
        for(int i=0;i<=MAXARRAY-1;i++){
            if(a[i]>b[i]){//if a is bigger in size
                for(int j = MAXARRAY-1;j >= 0;j--){//start from the back of the array
                    if((a[j]+carry)>=b[j]){//we can just subtract if the top item is larger
                        n[j]=a[j]-b[j]+carry;//subtract and include the carry
                        carry =0;//carry must be 0
                    }
                    else{//if i need to carry and b is bigger than a
                        a[j]=a[j]+10;//add 10 to the top function is equivalent to carrying
                        n[j]=a[j]-b[j]+carry;//then you may subtract
                        carry = -1;//carry a -1 if need to be carried
                        
                    }
                }
                break;//escape the i for loop
            }
            
            
            if(b[i]>a[i]){//if b is bigger in size
                for(int j = MAXARRAY-1;j >= 0;j--){//start from the back of the array
                    if((b[j]+carry)>=a[j]){//subtract the bigger number by smaller one and set to negative
                        n[j]=b[j]-a[j]+carry;//subtract and include the carry
                        carry =0;//carry must be 0
                    }
                    else{//if i need to carry and b is bigger than a
                        b[j]=b[j]+10;//add 10 to the top function is equivalent to carrying
                        n[j]=b[j]-a[j]+carry;//then you may subtract
                        carry = -1;//carry a -1 if need to be carried
                        
                    }
                }
                break;//escape the i for loop
            }
        }
        return n;
    }
    apint subtraction(apint first, apint second){//subracts 2 apints
        int[] a = first.getNumber();
        int[] b = second.getNumber();
        apint difference = new apint();//used to store the difference of the 2 apints
        int[] n = difference.getNumber();
        int carry=0;
        for(int i=0;i<=MAXARRAY-1;i++){
            if(a[i]>b[i]){//if a is bigger in size
                difference.negative=false;//then difference must not be negative
                for(int j = MAXARRAY-1;j >= 0;j--){//start from the back of the array
                    if((a[j]+carry)>=b[j]){//we can just subtract if the top item is larger
                        n[j]=a[j]-b[j]+carry;//subtract and include the carry
                        carry =0;//carry must be 0
                    }
                    else{//if i need to carry and b is bigger than a
                        a[j]=a[j]+10;//add 10 to the top function is equivalent to carrying
                        n[j]=a[j]-b[j]+carry;//then you may subtract
                        carry = -1;//carry a -1 if need to be carried
                        
                    }
                }
                break;//escape the i for loop
            }
            
            
            if(b[i]>a[i]){//if b is bigger in size
                difference.negative = true;//then difference must be negative
                for(int j = MAXARRAY-1;j >= 0;j--){//start from the back of the array
                    if((b[j]+carry)>=a[j]){//subtract the bigger number by smaller one and set to negative
                        n[j]=b[j]-a[j]+carry;//subtract and include the carry
                        carry =0;//carry must be 0
                    }
                    else{//if i need to carry and a is bigger than b
                        b[j]=b[j]+10;//add 10 to the larger function is equivalent to carrying
                        n[j]=b[j]-a[j]+carry;//then you may subtract
                        carry = -1;//carry a -1 if need to be carried
                        
                    }
                }
                break;//escape the i for loop
            }
        }
        return difference;
    }
    int[] shift(int[] a, int b){//used to shift an int array by b elements to b 0s at the end of the array
        int[] temp = new int[MAXARRAY];
        for (int i= MAXARRAY-1;i>=0+b;i--){// start at the back of the array and stop when there are b zeros at the end
            temp[i-b] = a[i];//shifts the data in the temp array
        }
        return temp; //return out shifted data
    }
    int[] append(int[] a, int b){//used to add int b at the end of the list
        int[] temp = new int[MAXARRAY];
        for (int i= MAXARRAY-1;i>=0;i--){//start from the back of the list
            temp[i-1] = a[i];//shift the data in the array of by one
        }
        temp[MAXARRAY-1]=b;// set the last element in the array to b
        return temp;//return our appended data
    }
    apint divide(apint dividend, apint divisor){//solved by counting the amount of subtractions
        int[] a = dividend.getNumber();
        int[] b = divisor.getNumber();
        apint quotient = new apint();//allocate memory for our quotient
        int[] n = quotient.getNumber();
        apint adder = new apint();//used to add or subtract by one
        adder.number[MAXARRAY-1]=1;
        if(dividend.Negative()==divisor.Negative()){//a test for negative a number
            
        }else{
            quotient.negative = true;
        }
        if(dividend.Negative()){//check to see if the dividend is negative or not
            while(dividend.Negative()){//while the dividend is negative
                
                dividend=subtraction(dividend,divisor);//subtract the dividend and divisor
                quotient=addition(quotient,adder);//add one to the quotient since there was a successful subtraction
                for (int i =0;i<=MAXARRAY-1;i++){//check to see if the dividend = 0
                    if(dividend.number[i] != 0){//continue in the while if dividend is not 0
                        break;
                    }else if(i==MAXARRAY-1){//case where it is a perfect divide
                        break;
                        
                    }else if(!dividend.Negative()){// case where there dividend is smaller than the divisor
                        quotient=subtraction(quotient,adder);//subtract one to get the lower bound
                        i=MAXARRAY-1;//need to leave this for loop
                    }
                }
            }
            
        }else if(!dividend.Negative()){//check to see if the dividend is negative or not
            while(!dividend.Negative()){//while the dividend is negative
                
                dividend=subtraction(dividend,divisor);//subtract the dividend and divisor
                quotient=addition(quotient,adder);//add one to the quotient since there was a successful subtraction
                for (int i =0;i<=MAXARRAY-1;i++){//check to see if the dividend = 0
                    if(dividend.number[i] != 0){//continue in the while if dividend is not 0
                        break;
                    }else if(i==MAXARRAY-1){//case where it is a perfect divide
                        break;
                    }else if(dividend.Negative()){// case where there dividend is smaller than the divisor
                        quotient=subtraction(quotient,adder);//subtract one to get the lower bound
                        i=MAXARRAY-1;//need to leave this for loop

                    }
                }
            }
            
        }
        return quotient;
    }
    int[] multi(int[]a,int[]b){//like the multiply fuction but its used in my factorial function
        int[] n = new int[MAXARRAY];
        int carry=0;
        int counter =0;
        int[] total = new int[MAXARRAY];
        for(int j=MAXARRAY-1;j>=0;j--){
            for(int k=MAXARRAY-1;k>=0;k--){
                total[k]=(a[k]*b[j]+carry);//multipy a and b and add carry
                carry = total[k]/10;//carry is total over 10 for example if total is 54 then carry is 5
                total[k]= total[k]%10;//mod by 10 to get rid of the higher order terms
            }
            total=shift(total,MAXARRAY-1-j);//method to know how much to shift data to the left
            n= add(total,n);//add the total of the mulitpy and n is the running total of the whole process
        }
        return n;
    }
    apint multiply(apint first,apint second){
        int[] a = first.getNumber();
        int[] b = second.getNumber();
        apint product = new apint();
        int[] n = product.getNumber();
        int carry=0;
        int counter =0;
        int[] total = new int[MAXARRAY];
        if(first.Negative()==second.Negative()){//test to figure out if product is negative or not
            
        }else{
            product.negative = true;
        }
        for(int i=0;i<=MAXARRAY-1;i++){//loop to figure out which operation goes on top and which goes on the bottom
            if(a[i]>b[i]){//if the first apint element is larger
                for(int j=MAXARRAY-1;j>=0;j--){
                    for(int k=MAXARRAY-1;k>=0;k--){
                        total[k]=(a[k]*b[j]+carry);//multipy a and b and add carry
                        carry = total[k]/10;//carry is total over 10 for example if total is 54 then carry is 5
                        total[k]= total[k]%10;//mod by 10 to get rid of the higher order terms
                    }
                    total=shift(total,MAXARRAY-1-j);//method to know how much to shift data to the left
                    n= add(total,n);//add the total of the mulitpy and n is the running total of the whole process
                }
                product.number = n;
                return product;
            }
            if(b[i]>a[i]){//if the second apint element is larger
                for(int j=MAXARRAY-1;j>=0;j--){
                    for(int k=MAXARRAY-1;k>=0;k--){
                        total[k]=(a[k]*b[j]+carry);//multipy a and b and add carry
                        carry = total[k]/10;//carry is total over 10 for example if total is 54 then carry is 5
                        total[k]= total[k]%10;//mod by 10 to get rid of the higher order terms
                    }
                    total=shift(total,MAXARRAY-1-j);//method to know how much to shift data to the left
                    n= add(total,n);//add the total of the mulitpy and n is the running total of the whole process
                }
                product.number = n;
                return product;
            }
        }
        return product;
    }
    apint factorial(int one){//method to handle 1000! and passes in an int because factorial wont happen with numbers bigger than the int max
        apint alpha = new apint(one);//
        int[] a = alpha.number;
        int[] b = new int[MAXARRAY];
        b[MAXARRAY-1]=1;//used to decrement one to figure out when to stop multiplying

        apint tot = new apint(one);//total starts at one
        int[] total = tot.number;
        
        for(int i=one-2;i>=0;i--){//subtract by 2 becuase you dont need to multiply by one at the end of the factorial function
            a = subtract(a,b);//subtract the count by one
            total = multi(a,total);//multiply the running total and the number in a we are at.
            
        }
        apint fact = new apint(total);// make an apint with total stored in it
        
        return fact;//return the product of the factorial function
    }

}

File: CMPS101-Algorithms and Abstract Data Types/Programming Assignments/CMPS101S18PA2/apint.c
/*
 CREATED: Derrick DeBose April 27, 2018
 CHANGED: May 2, 2018
 
 apint.c is the file that is the apint ADT that creates apints
 from strings or ints and can manipulate the data with methods to add,
 subtract, and multiply the numbers. There is a method that can print out the apint to the system

*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "apint.h"
#define MAXARRAY 100
#define TRUE 1
#define FALSE 0

/*
 This struct is the Arbitrary Precession Integer
 */
typedef struct newAPINTStruct{
    int *number;//array of ints to represent an apint
    int boolean;//negative flag
}newAPINTStruct;

/*
 Method accepts a pointer an apint struct and allocates memory for the apint
 */
void newApint(apint *a){// a pointer to the struct
    (*a)= calloc(1,sizeof(apint));//allocate memory for the apint pointer
    (*a)->number= calloc(MAXARRAY,sizeof(int));//allocate memory for the array
    for(int j=0;j<MAXARRAY;j++){
        (*a)->number[j]=0;//set the array to zero
    }
    
    (*a)->boolean = FALSE;//set the negative flag to false
}

/*
 Method accepts a pointer an apint struct and a string of characters that will be put into the apint structure
 */
void newApintStr(apint *api, char c[]) {//converts a string of ints to an apint
    int length = strlen(c);//length of the string
    for(int k=0;k<length-1;k++){
        if(c[k]=='-'){//check to see if the number is negative
            (*api)->boolean= TRUE;
            
        }
    }
    int counter = MAXARRAY-1;//start from the back of the int array
    for( int i=length-1;i>=0;i--){//start at the end of the string
        if(c[i]=='-'){//check for a negative sign
            (*api)->number[counter] =(c[i]-'-');
        }
        
        else if(c[i]=='+'){//check for a positive sign
            (*api)->number[counter] =(c[i]-'+');
        }
        
        else{//converts the char to a number and stores it in the array
            (*api)->number[counter] =(c[i]-'0');
        }
        
        counter--;
    }
}

/*
 Method accepts a pointer an apint struct and an int that will be put into the apint structure
 */
void newApintInt(apint *api, int a) {
    char c[MAXARRAY];//allocates memory for a string
    sprintf(c,"%d",a);//method converts an int to a string
    int length= strlen(c);
    for(int k=0;k<=MAXARRAY-1;k++){
        if(c[k]=='-'){//check to see if the number is negative
            (*api)->boolean= TRUE;
        }
    }
    int counter = MAXARRAY-1;
    for( int i=length-1;i>=0;i--){
        if(c[i]=='-'){//check for a negative sign
            (*api)->number[counter] =(c[i]-'-');
        }
        
        else if(c[i]=='+'){//check for a positive sign
            (*api)->number[counter] =(c[i]-'+');
        }
        
        else{//converts the char to a number and stores it in the array
            (*api)->number[counter] =(c[i]-'0');
            
        }
        counter--;
    }
}

/*
The print method checks if the apint passed in is
zero and will print out zero otherwise prints out the number and removes all the leading zeros in the apint.
*/
void print(apint a){//method used to print out an apint
    if(a->boolean == TRUE){//if the number is negative
        printf("-");
    }
    int counter =0;
    while(a->number[counter]==0){//checks ends when has passed all the leading zeros
        counter++;
        if(counter == MAXARRAY){//if made it to the max array then there is a 0 stored
            printf("0");
        }
    }
    for(int i=0+counter;i<MAXARRAY; i++){//starts from counter to maxarray
        printf("%d", a->number[i]);;//add number to the back of string
    }
    printf("\n\n");
}

/*The add method in my code starts at the back of the list we add the 2 ints
 and then calculate carry based on the sum. The result stores the sum.
 The arguments are a+b=result
*/
void add(apint a, apint b, apint result){//adds 2 apints and store the sum in result
    int carry =0;//used to determine what needs to be carried
    int total =0;//used as a temp to determine int needs to be stored in the array
    for(int i=MAXARRAY-1;i>=0;i--){//start at the back of the list to add
        total = (a->number[i])+(b->number[i])+carry;//total is the addition of the 2 numbers including the carry
        (result->number[i])=total%10;//if total is  15 or 25, result=5
        total=total/10;//if total is 15 or 25, total 1,2
        carry= total;
    }
}

/*
 The subtract method in my code first determines which number is bigger in size.
 If the second number is larger than the first then we will subtract the second number from the first
 and set the difference to negative. We subtract a-b when a>b and b-a when b>a.
 The difference is a-b = result
*/
void subtract(apint a, apint b, apint result){
    int carry =0;
    for(int i=0;i<=MAXARRAY-1;i++){//loop to determine which number is larger
        if((a->number[i])>(b->number[i])){//if a is bigger in size
            result->boolean=FALSE;//then difference must not be negative
            for(int j = MAXARRAY-1;j >= 0;j--){//start from the back of the array
                if((a->number[j]+carry)>=(b->number[j])){//we can just subtract if the top item is larger
                    (result->number[j])=a->number[j]-(b->number[j])+carry;//subtract and include the carry
                    carry =0;//carry must be 0
                }
                else{//if i need to carry and b is bigger than a
                    a->number[j]=a->number[j]+10;//add 10 to the top function is equivalent to carrying
                    result->number[j]=a->number[j]-b->number[j]+carry;//then you may subtract
                    carry = -1;//carry a -1 if need to be carried
                    
                }
            }
            break;//escape the i for loop
        }
        if((b->number[i])>(a->number[i])){//if b is bigger in size
            result->boolean = TRUE;//then difference must be negative
            for(int j = MAXARRAY-1;j >= 0;j--){//start from the back of the array
                if((b->number[j]+carry)>=a->number[j]){//subtract the bigger number by smaller one and set to negative
                    result->number[j]=b->number[j]-a->number[j]+carry;//subtract and include the carry
                    carry =0;//carry must be 0
                }
                else{//if i need to carry and a is bigger than b
                    b->number[j]=b->number[j]+10;//add 10 to the larger function is equivalent to carrying
                    result->number[j]=b->number[j]-a->number[j]+carry;//then you may subtract
                    carry = -1;//carry a -1 if need to be carried
                    
                }
            }
            break;//escape the i for loop
        }
    
    }
}

/*
The shift method to shift the data in order for there to be zeros at the end to hold multiplication process.
Method used in the multiply function to add b zeros to the end of the apint.
*/
apint shift(apint a, int b){
    apint total;//a temp int to transfer the data
    newApint(&total);
    for(int i=MAXARRAY-1;i>=0+b;i--){
        total->number[i-b]=a->number[i];//more the data b spaces to the left
    }
    return total;
}

/*
The multiply method first determines which number is bigger in size.
Total is calculated by multiplying the 2 apints and adding by the carry at that index then uses a shift method
to shift the data in order for there to be zeros at the end to hold the long multiplication process.
The arguments passed in are a x b = result and the total apint is an
empty apint that is used to be passed into the shift method
*/
void multiply(apint a, apint b, apint total, apint result){
    int carry=0;
    if(a->boolean==b->boolean){//test to figure out if product is negative or not
        result->boolean=FALSE;
    }else{
        result->boolean = TRUE;
    }
    for(int i=0;i<=MAXARRAY-1;i++){//loop to figure out which operation goes on top and which goes on the bottom
        if((a->number[i])>(b->number[i])){//if the first apint element is larger
            for(int j=MAXARRAY-1;j>=0;j--){
                carry=0;
                for(int k=MAXARRAY-1;k>=0;k--){
                    (total->number[k])=(a->number[k])*(b->number[j])+carry;//multipy a and b and add carry
                    carry = (total->number[k])/10;//carry is total over 10 for example if total is 54 then carry is 5
                    (total->number[k])= (total->number[k])%10;//mod by 10 to get rid of the higher order terms
                }
                total = shift(total,MAXARRAY-1-j);//method to know how much to shift data to the left
                add(total,result,result);//add the total of the mulitpy and n is the running total of the whole process
            }
            break;
        }
        if((b->number[i])>(a->number[i])){//if the second apint element is larger
            for(int j=MAXARRAY-1;j>=0;j--){
                carry=0;
                for(int k=MAXARRAY-1;k>=0;k--){
                    (total->number[k])=(a->number[k]*b->number[j]+carry);//multipy a and b and add carry
                    carry = (total->number[k])/10;//carry is total over 10 for example if total is 54 then carry is 5
                    (total->number[k])= (total->number[k])%10;//mod by 10 to get rid of the higher order terms
                }
                total =shift(total,MAXARRAY-1-j);//method to know how much to shift data to the left
                add(total,result,result);//add the total of the mulitpy and n is the running total of the whole process
            }
            break;
        }
    }
}



